diff --git a/linux-5.14.4/arch/x86/entry/syscalls/syscall_64.tbl b/linux-5.14.4/arch/x86/entry/syscalls/syscall_64.tbl
index f6b57799c..d435c9c17 100644
--- a/linux-5.14.4/arch/x86/entry/syscalls/syscall_64.tbl
+++ b/linux-5.14.4/arch/x86/entry/syscalls/syscall_64.tbl
@@ -368,8 +368,10 @@
 444	common	landlock_create_ruleset	sys_landlock_create_ruleset
 445	common	landlock_add_rule	sys_landlock_add_rule
 446	common	landlock_restrict_self	sys_landlock_restrict_self
-447	common	memfd_secret		sys_memfd_secret
-
+447	common	memfd_secret        sys_memfd_secret
+448 common  hello_world         sys_hello_world
+449 common  kernel_2d_memcpy    sys_kernel_2d_memcpy
+450 common  delay_sched_process sys_delay_sched_process
 #
 # Due to a historical design error, certain syscalls are numbered differently
 # in x32 as compared to native x86_64.  These syscalls have numbers 512-547.
diff --git a/linux-5.14.4/include/linux/sched.h b/linux-5.14.4/include/linux/sched.h
index ec8d07d88..f7e6585f4 100644
--- a/linux-5.14.4/include/linux/sched.h
+++ b/linux-5.14.4/include/linux/sched.h
@@ -475,6 +475,9 @@ struct sched_entity {
 	u64				prev_sum_exec_runtime;
 
 	u64				nr_migrations;
+	u64				log_timestamp_flag;
+	u64				pid_of_process;
+	u64				delay_ms;
 
 	struct sched_statistics		statistics;
 
diff --git a/linux-5.14.4/include/linux/syscalls.h b/linux-5.14.4/include/linux/syscalls.h
index 69c9a7010..5e68dcb30 100644
--- a/linux-5.14.4/include/linux/syscalls.h
+++ b/linux-5.14.4/include/linux/syscalls.h
@@ -1378,4 +1378,5 @@ int __sys_getsockopt(int fd, int level, int optname, char __user *optval,
 		int __user *optlen);
 int __sys_setsockopt(int fd, int level, int optname, char __user *optval,
 		int optlen);
+asmlinkage long sys_delay_sched_process(unsigned long pid_of_process,unsigned long delay_ms);
 #endif
diff --git a/linux-5.14.4/kernel/sched/core.c b/linux-5.14.4/kernel/sched/core.c
index a2403432f..8325136f0 100644
--- a/linux-5.14.4/kernel/sched/core.c
+++ b/linux-5.14.4/kernel/sched/core.c
@@ -3956,6 +3956,9 @@ static void __sched_fork(unsigned long clone_flags, struct task_struct *p)
 	p->se.prev_sum_exec_runtime	= 0;
 	p->se.nr_migrations		= 0;
 	p->se.vruntime			= 0;
+	p->se.log_timestamp_flag = 0;
+	p->se.pid_of_process = 0;
+	p->se.delay_ms = 0;
 	INIT_LIST_HEAD(&p->se.group_node);
 
 #ifdef CONFIG_FAIR_GROUP_SCHED
diff --git a/linux-5.14.4/kernel/sched/fair.c b/linux-5.14.4/kernel/sched/fair.c
index 30a6984a5..9cdf7eaa0 100644
--- a/linux-5.14.4/kernel/sched/fair.c
+++ b/linux-5.14.4/kernel/sched/fair.c
@@ -811,8 +811,14 @@ static void update_curr(struct cfs_rq *cfs_rq)
 	curr->sum_exec_runtime += delta_exec;
 	schedstat_add(cfs_rq->exec_clock, delta_exec);
 
-	curr->vruntime += calc_delta_fair(delta_exec, curr);
-	update_min_vruntime(cfs_rq);
+	if(curr->delay_ms!=0){
+		curr->vruntime += calc_delta_fair(delta_exec, curr)+(curr->delay_ms*1000000);
+		update_min_vruntime(cfs_rq);
+	}
+	else{
+		curr->vruntime += calc_delta_fair(delta_exec, curr);
+		update_min_vruntime(cfs_rq);
+	}
 
 	if (entity_is_task(curr)) {
 		struct task_struct *curtask = task_of(curr);
@@ -4534,6 +4540,12 @@ pick_next_entity(struct cfs_rq *cfs_rq, struct sched_entity *curr)
 		se = cfs_rq->last;
 	}
 
+	if(se->log_timestamp_flag!=0){
+		s64 now = ktime_to_ns(ktime_get_real());
+		u64 pid_of_process_temp = se->pid_of_process; 
+		printk(KERN_INFO "Scheduled process with PID %lu at time: %ld (ns)\n", pid_of_process_temp, now);
+	}
+
 	return se;
 }
 
diff --git a/linux-5.14.4/kernel/sys.c b/linux-5.14.4/kernel/sys.c
index ef1a78f5d..2a0b2f8e5 100644
--- a/linux-5.14.4/kernel/sys.c
+++ b/linux-5.14.4/kernel/sys.c
@@ -41,6 +41,7 @@
 #include <linux/syscore_ops.h>
 #include <linux/version.h>
 #include <linux/ctype.h>
+#include <linux/module.h>
 #include <linux/syscall_user_dispatch.h>
 
 #include <linux/compat.h>
@@ -200,6 +201,81 @@ static int set_one_prio(struct task_struct *p, int niceval, int error)
 	return error;
 }
 
+SYSCALL_DEFINE1(hello_world, char*, msg)
+{
+	char buf[265];
+	long copied = strncpy_from_user(buf, msg, sizeof(buf));
+	if(copied<0 || copied == sizeof(buf)){
+		return -EFAULT;
+	}
+	printk(KERN_INFO "hello_world called with\"%s\"\n", buf);
+	return 0;
+}
+
+SYSCALL_DEFINE3(kernel_2d_memcpy, float*, src, float*, dest, unsigned long, myArrSize)
+{
+	float* myArr = (float*)kmalloc(myArrSize, GFP_KERNEL);
+	int my_ret_val = copy_from_user(myArr, src, myArrSize);
+
+	if(!myArr){
+		return -ENOMEM;
+	}
+
+	if(!access_ok(src, myArrSize)){
+		kfree(myArr);
+		return -EFAULT;
+	}
+	
+	if(!access_ok(dest, myArrSize)){
+		kfree(myArr);
+		return -EFAULT;
+	}
+
+	if(my_ret_val>0){
+		kfree(myArr);
+		return -EFAULT;
+	}
+
+	my_ret_val = copy_to_user(dest, myArr, myArrSize);
+
+	if(my_ret_val>0){
+		kfree(myArr);
+		return -EFAULT;
+	}
+
+	kfree(myArr);
+	return 0;
+}
+
+SYSCALL_DEFINE2(delay_sched_process,unsigned long,pid_of_process,unsigned long,delay_ms)
+{
+	if(delay_ms<0 || pid_of_process<0){
+		printk(KERN_INFO "Invalid value of delay time given!\n");
+		return -EFAULT;
+	}
+	
+	int myFlag = 0;
+	struct task_struct *p;
+	
+	for_each_process(p){
+		if((unsigned long)p->pid==pid_of_process){
+			p->se.pid_of_process = pid_of_process;
+			p->se.delay_ms = delay_ms;
+			p->se.log_timestamp_flag = 1;
+			printk(KERN_INFO "Fields has been updated for PID: %lu \n", pid_of_process);
+			myFlag = 1;
+		}
+	}
+	
+	if(myFlag == 0){
+		printk(KERN_INFO "No process with the given PID found!\n");
+		return -EFAULT;
+	}
+	
+	printk(KERN_INFO "System call finished!\n");
+	return 0;
+}
+
 SYSCALL_DEFINE3(setpriority, int, which, int, who, int, niceval)
 {
 	struct task_struct *g, *p;
